antibody.go

```diff
@@ -28,29 +37,91 @@ func New(home string, r io.Reader) *Antibody {
 
 // Bundle processes all given lines and returns the shell content to execute
 func (a *Antibody) Bundle() (result string, err error) {
-	var g errgroup.Group
-	var lock sync.Mutex
-	var shs []string
-	scanner := bufio.NewScanner(a.r)
-	for scanner.Scan() {
-		l := scanner.Text()
-		g.Go(func() error {
-			l = strings.TrimSpace(l)
-			if l != "" && l[0] != '#' {
-				s, err := bundle.New(a.Home, l).Get()
-				lock.Lock()
-				shs = append(shs, s)
-				lock.Unlock()
-				return err
+	file := a.r
+
+	input_lines := make(chan Result)
+	results := make(chan Result)
+
+	// I think we need a wait group, not sure.
+	wg := new(sync.WaitGroup)
+
+	// workers
+	for w := 1; w <= 8; w++ {
+		wg.Add(1)
+		go func() {
+			// Decreasing internal counter for wait-group as soon as goroutine finishes
+			defer wg.Done()
+
+			for res := range input_lines {
+				log.Debugf("Bundling: %s", res.line)
+				res.line = strings.TrimSpace(res.line)
+
+				if res.line == "" || res.line[0] == '#' {
+					continue
+				}
+
+				val, err := bundle.New(a.Home, res.line).Get()
+				res.line = val
+
+				if err != nil {
+					log.Fatalf("Error processing bundle=%s: %s", res.line, err)
+				} else {
+					results <- res
+				}
 			}
-			return nil
-		})
+		}()
+	}
+
+	// source
+	go func() {
+		log.Debugf("Reading bundles")
+
+		scan := bufio.NewScanner(file)
+
+		idx := 0
+		for scan.Scan() {
+			line := scan.Text()
+			line = strings.TrimSpace(line)
+
+			input_lines <- Result{idx, line}
+			idx++
+		}
+
+		err := scan.Err()
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		close(input_lines)
+
+		log.Debugf("Done reading bundles")
+	}()
+
+	// waiter
+	go func() {
+		wg.Wait()
+		close(results)
+	}()
+
+	// collect
+	var all_results []Result
```

Comment:
Ah, of course, how else am I supposed to know it's a variable and not a function! Oh, right, I can't because go's formatting is broken.